### PR TITLE
[MLIR] Make `OneShotModuleBufferize` use `OpInterface`

### PR DESCRIPTION
**Pull request by tzunghanjuang**
**Description:** 

`OneShotModuleBufferize` deals with the bufferization of `FuncOp`, `CallOp` and `ReturnOp` but they are hard-coded. Any custom function-like operations will not be handled. The PR replaces a part of `FuncOp` and  `CallOp` with `FunctionOpInterface` and `CallOpInterface` in `OneShotModuleBufferize` so that custom function ops and call ops can be bufferized.

**Related Discord Discussion:** [Link](https://discord.com/channels/636084430946959380/642426447167881246/1280556809911799900)

### CODE REVIEW

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp [hunk 1]
```diff
@@ -247,6 +247,15 @@ static func::FuncOp getCalledFunction(func::CallOp callOp) {
       SymbolTable::lookupNearestSymbolFrom(callOp, sym));
 }
 
+static FunctionOpInterface getCalledFunction(CallOpInterface callOp) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:250]: I think the above `static func::FuncOp getCalledFunction(func::CallOp callOp)` can be deleted.

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp [hunk 2]
```diff
@@ -388,27 +401,33 @@ mlir::bufferization::analyzeModuleOp(ModuleOp moduleOp,
     return failure();
 
   // Analyze ops.
-  for (func::FuncOp funcOp : orderedFuncOps) {
-    if (!state.getOptions().isOpAllowed(funcOp))
+  for (FunctionOpInterface funcOp : orderedFuncOps) {
+
+    // The following analysis is specific to the FuncOp type.
+    if(!isa<FuncOp>(funcOp))
+        continue;
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:408]: I think this is functionally correct but could be a performance problem in practice. This piece of code is gathering information about the relationship between func args and func results. This is needed to compute aliasing values during `CallOpInterface::getAliasingValues`. For functions that were not analyzed, we assume that all CallOp operands potentially alias will all CallOp results. You may get additional buffer copies because of that. Could that become a problem?

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp [hunk 3]
```diff
@@ -291,26 +300,30 @@ static bool hasTensorSignature(func::FuncOp funcOp) {
...
-    if (!funcOp.getBody().empty()) {
-      func::ReturnOp returnOp = getAssumedUniqueReturnOp(funcOp);
-      if (!returnOp)
-        return funcOp->emitError()
-               << "cannot bufferize a FuncOp with tensors and "
-                  "without a unique ReturnOp";
+  DenseMap<FunctionOpInterface, unsigned> numberCallOpsContainedInFuncOp;
+  WalkResult res = moduleOp.walk([&](FunctionOpInterface funcOp) -> WalkResult {
+    // Only handle ReturnOp if funcOp is exactly the FuncOp type.
+    if(isa<FuncOp>(funcOp)) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:312]: I think you can count the number of `ReturnLike` (that's a trait) operations in the the region of the `funcOp`. (Not in nested regions.)

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 2 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:408]: Yes, I can see it is bad for the performance. This part involves `FuncAnalysisState` from `FuncBufferizableOpInterfaceImpl`, which is also hard-coded with `FuncOp`. I will try replacing it with `FunctionOpInterface` as well.

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 1 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:242]: Removed. Thank you.

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 3 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: Hi, we have made `getAssumedUniqueReturnOp` to detect `ReturnLike` ops and made it return `funcOp` if there is no `ReturnLike` op. When it returns `funcOp` ifselt, the analysis in `aliasingFuncOpBBArgsAnalysis` and `foldMemRefCasts` will be skipped.

In summary, we think there are 3 scenarios:
- Exaclty one `ReturnLike` op => Perform the return-related analysis.
- No `ReturnLike` op  => Skip the analysis. No error is thrown.
- Multiple `ReturnLike` ops => Throw an error.

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 2 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: We have made all fields in `FuncAnalysisState` use `FunctionOpInterface` instead of `func::FuncOp`.

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: @tzunghanjuang is this because the functions have no body (i.e., they are just a declaration)? Or do we have an terminator op that does not implement the ReturnLike trait that is triggering this?

Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: For a bit more context, we think there are some `FunctionOpInterface`s ,like [`transform.NamedSequenceOp`](https://github.com/llvm/llvm-project/blob/109cd11dc4aea6b3596f8b2cb5a719f35b190cfa/mlir/include/mlir/Dialect/Transform/IR/TransformOps.td#L993-L997), having no `ReturnLike` at the end.

Example: https://github.com/llvm/llvm-project/blob/main/mlir/test/Dialect/Vector/transform-vector.mlir

Comment (matthias-springer) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: I would throw an error even if there no `ReturnLike` op, and relax this requirement only if needed. `transform.named_sequence` does not return tensors, so we wouldn't bufferize it in the first place.

**Review by matthias-springer — APPROVED** [review]

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp [hunk 4]
```diff
@@ -349,11 +354,15 @@ getFuncOpsOrderedByCalls(ModuleOp moduleOp,
...
 /// the return type of the function.
-static void foldMemRefCasts(func::FuncOp funcOp) {
-  if (funcOp.getBody().empty())
+static void foldMemRefCasts(FunctionOpInterface funcOp) {
+  if (funcOp.getFunctionBody().empty())
+    return;
+
+  Operation *returnOp = getAssumedUniqueReturnOp(funcOp);
+
+  if (!returnOp || returnOp == funcOp)
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:363]: Why is this needed? Why would the return op be the same op as the function op?

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 3 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: Thank you for the suggestion. We have updated `getAssumedUniqueReturnOp` so that it no longer returns `funcOp`. We did not find a directly determine if a `FunctionOpInterface` is shipped with any `ReturnLike` Op, so we used  `funcOp.getNumResults()` to check if there is no return type. 

With this method, we still have to keep the check near `aliasingFuncOpBBArgsAnalysis` and `foldMemRefCasts` or the following example will fail. The example uses `FuncOP` and returns nothing. If we check `funcOp.getNumResults()` too early, the bufferization of the function's body will be entirely skipped and fail the file check. 

https://github.com/llvm/llvm-project/blob/fc7a8936208847b66df9b5f93ba2222b93b9f8b6/mlir/test/Dialect/ControlFlow/one-shot-bufferize-analysis.mlir#L91-L111

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 4 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: This is to prevent `NamedSequenceOp` from being analyzed because it has no return. We have updated  `getAssumedUniqueReturnOp` so the guard is moved to the start of `aliasingFuncOpBBArgsAnalysis` (by checking the number of return types).

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 3 above)
Comment (erick-xanadu) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: Hi @tzunghanjuang, you should not use `getNumResults` in this manner:

```
    if (!funcOp.getFunctionBody().empty() && funcOp.getNumResults() != 0) {
```

One of the errors appears to be in the tests themselves. I'll explain here for @matthias-springer:

The tests that involve the transform interpreter will attempt to bufferize the root module which contains the `transform::NamedSequenceOp`. The `transform::NamedSequenceOp` implements `FunctionOpInterface`, but the dialect does not have a `ReturnLike` op. The one-shot bufferizer will attempt to bufferize all nested operations inside the root module which includes the `transform::NamedSequenceOp`. However, this will obviously fail.

So I believe the correct way to solve this is to change the test such that the bufferization only happens in the intended sections of the module. This looks like it can be achieved by specifying a payload target with the option `--debug-payload-root-tag=<string>` or (as I just looked into other tests) using the `transform.structured.match` operation.

Looking at the tests, it seems that `transform.structured.match` is preferred, but I'll let others chime in.

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp [hunk 5]
```diff
@@ -349,11 +348,11 @@ getFuncOpsOrderedByCalls(ModuleOp moduleOp,
 /// most generic layout map as function return types. After bufferizing the
 /// entire function body, a more concise memref type can potentially be used for
 /// the return type of the function.
-static void foldMemRefCasts(func::FuncOp funcOp) {
-  if (funcOp.getBody().empty())
+static void foldMemRefCasts(FunctionOpInterface funcOp) {
+  if (funcOp.getFunctionBody().empty() || funcOp.getNumResults() == 0)
```
Comment (erick-xanadu) [inline mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:352]: I think if this breaks some tests (like in the case of the `transform::NamedSequenceOp` being bufferized) then the tests should be changed.

```suggestion
  if (funcOp.getFunctionBody().empty())
```

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp [hunk 6]
```diff
@@ -126,16 +127,15 @@ static void annotateEquivalentReturnBbArg(OpOperand &returnVal,
 /// Store function BlockArguments that are equivalent to/aliasing a returned
 /// value in FuncAnalysisState.
 static LogicalResult
-aliasingFuncOpBBArgsAnalysis(FuncOp funcOp, OneShotAnalysisState &state,
+aliasingFuncOpBBArgsAnalysis(FunctionOpInterface funcOp, OneShotAnalysisState &state,
                              FuncAnalysisState &funcState) {
-  if (funcOp.getBody().empty()) {
+  if (funcOp.getFunctionBody().empty() || funcOp.getNumResults() == 0) {
```
Comment (erick-xanadu) [inline mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp:132]: Same here.

```suggestion
  if (funcOp.getFunctionBody().empty()) {
```

File: mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp (hunk 3 above)
Comment (tzunghanjuang) [reply mlir/lib/Dialect/Bufferization/Transforms/OneShotModuleBufferize.cpp]: Thank you for the clarification. I have updated the `transform.named_sequence` tests with `-debug-payload-root-tag` option. The previous workaround has also been removed. The "no ReturnLike op" cases will throw errors.

File: mlir/test/Examples/transform/ChH/full.mlir [hunk 7]
```diff
@@ -19,6 +19,7 @@
 // tensors annotated with attributes from the `bufferization` dialect. These
 // attributes hint the bufferization pass to assume buffers can be directly
 // used for these tensors without reshaping.
+module @payload attributes { transform.target_tag = "payload" } {
 func.func @conv(
     %input: !tinput {bufferization.writable = false,
                      bufferization.access = "read",
```
Comment (tzunghanjuang) [inline mlir/test/Examples/transform/ChH/full.mlir:25]: Hi @matthias-springer , we are trying to use `debug-payload-root-tag=payload` to limit the `moduleOP bufferization` to the payload module. (Or it will trigger an error for no `ReturnLike`.)

It seems that the added payload module in this test blocks some passes after `transform-interpreter`.  The checks fail since the IRs like `llvm.intr.fma` are not produced. Maybe there are some integration issues? Is there any better way to exclude `transform.named_sequence` from bufferziation? Thank you very much.
 
https://github.com/llvm/llvm-project/blob/828783177f71d95522763fba92ef1e42dc6101c7/mlir/test/Examples/transform/ChH/full.mlir#L379-L388

Comment (matthias-springer) [reply mlir/test/Examples/transform/ChH/full.mlir:25]: I forgot how exactly `transform.target_tag = "payload"` works. Maybe the transform script is only applied to the contents of the module but not the module itself. Can you try adding a nested `module`?

Comment (tzunghanjuang) [reply mlir/test/Examples/transform/ChH/full.mlir:25]: Thank you. I have tried adding a nested module but we still got the same error. We will try exploring the possible fix.

Comment (erick-xanadu) [reply mlir/test/Examples/transform/ChH/full.mlir:25]: Hi @tzunghanjuang 

can you try the following patch

```patch
diff --git a/mlir/test/Examples/transform/ChH/full.mlir b/mlir/test/Examples/transform/ChH/full.mlir
index 005ac5a9ba..190d9672fd 100644
--- a/mlir/test/Examples/transform/ChH/full.mlir
+++ b/mlir/test/Examples/transform/ChH/full.mlir
@@ -1,8 +1,6 @@
 // RUN: mlir-opt %s --transform-interpreter="debug-payload-root-tag=payload" \
-// RUN:             --test-transform-dialect-erase-schedule \
-// RUN:             --math-uplift-to-fma \
-// RUN:             --convert-bufferization-to-memref \
-// RUN:             --test-lower-to-llvm |\
+// RUN:             --test-transform-dialect-erase-schedule \|
+// RUN: mlir-opt -pass-pipeline='builtin.module(builtin.module(math-uplift-to-fma,convert-bufferization-to-memref,test-lower-to-llvm))' - \|
 // RUN: FileCheck %s
 
 // Fixed-size tensor types to be used in convolution.
 ```
 
 I think the issue is not from the payload module not being transformed correctly by the transform schedule, but by the other passes not being schedule properly. So I created a pass pipeline that acts on nested modules. I tested this locally and it works.

Comment (tzunghanjuang) [reply mlir/test/Examples/transform/ChH/full.mlir:23]: Thank you @erick-xanadu. It also works on my side. I have pushed the patch to see what will happen on the CI.

**Comment by matthias-springer** [comment]
@tzunghanjuang This broke an integration test. Can you take fix it an reopen this PR, then I'm going to merge the fixed PR again.

**Comment by tzunghanjuang** [comment]
Hi @matthias-springer , the error is triggered by `CallIndirectOp`. Since the function called by `CallIndirectOp` cannot be known statically,  `getCalledFunction` will return `nullptr` and then break the assertion. We fixed it by making the function analysis skip `CallIndirectOp`. Do you have any suggestions for handling this case? Thank you.

**Comment by tzunghanjuang** [comment]
I do not have the access to reopen this PR. I create a new one with the fix: https://github.com/llvm/llvm-project/pull/110322.
